### PR TITLE
Process most 'require's at resolution time

### PR DESCRIPTION
**Pull request by bradcray**
Prior to this effort, 'require' statements were processed at parse-time.  This had a few consequences:  (1) It meant that their arguments had to be string literals rather than more general param string expressions; (2) it meant that its requirements would be added to the list regardless of whether or not the code in question was actually 'use'd.  With this change, we handle (most) require statements at function resolution time, permitting us to handle general param expressions and to only process those require statements that are in active code.

This change was motivated by a user who wanted to have a sub-module with its own requirements, in order to have the user be able to opt-in to not-generally-available files, and to have them only be implied if the sub-module is 'use'd.  In thinking about this and testing it, I realized that the early handling I'd originally implemented would cause the requirements to fire in any case.  I'd long intended to make require statement support more general param expressions, so this was a good chance to handle both things at the same time.

One case that I still handle early is 'require' statements of ".chpl" files because we need to process these at parse-time, since we don't currently have the ability to go back and selectively re-compile some additional '.chpl' files (and may never want to).  Needless to say, this  means that .chpl files must be specified with string literals rather than param string expressions.

The implementation approach here was to:

* Make the processStringInRequirementStmt() routine call-able at either parse- or resolution-time, to accept an additional argument indicating which case it was, to have it handle ".chpl" files early, and to have it handle other ones later.

* To introduce a new primitive (PRIM_REQUIRE) to maintain require statements in the AST through function resolution.

* To process PRIM_REQUIRE statements in the postFold().

* To have the list of file dependences check for duplicates.  This could've been useful earlier to catch user errors that we hadn't bothered guarding against; now it's necessary even for user code without duplicates at times, because the retention of the require statements can cause them to be cloned if, for example, they appear in a generic function that is cloned.  In one test, this caused a .h file to be #included twice, and because the header did not guard against this usage, it resulted in compile-time problems.  It seems plausible that the duplicate listing of a .c file could also run into problems.  In any case, it seems worthwhile to filter duplicates out to avoid providing redundant information to the C compiler.

Added new tests:

* requireChapelLiteral: ensures that a 'require' of a .chpl string literal works
* requireChapelParam: tests the error message for a 'param' .chpl requirement
* requireInSubMod: tests that we don't process requirements for un-'use'd sub-modules
* requireInTopMod: tests that we don't process requirements for un-'use'd top-level modules
* requireParamString*: checks to make sure that 'require'ments of param strings work

[Tested on linux64]

### CODE REVIEW

**Comment by bradcray** [comment]
@nspark : You may be interested in this one.

File: compiler/AST/build.cpp [hunk 1]
```diff
@@ -512,17 +527,27 @@ BlockStmt* buildRequireStmt(CallExpr* args) {
     Expr* useArg = expr->remove();
 
     //
-    // if this is a string argument to 'require', process it
+    // if this is a string literal, process it if we should
     //
     if (const char* str = toImmediateString(useArg)) {
-      processStringInRequireStmt(str);
+      if (processStringInRequireStmt(str)) {
```
Comment (vasslitvinov) [inline compiler/AST/build.cpp]: I suggest passing an explicit actual argument for the parseTime formal.

Comment (vasslitvinov) [inline compiler/AST/build.cpp]: ... and then remove the default value for that formal in the prototype decl.

Comment (bradcray) [inline compiler/AST/build.cpp]: I agree.  I thought of this after filing the PR, but then figured I'd wait and see what the review said first.  Thanks!

**Comment by lydia-duncan** [comment]
Does the presence of duplicate files "required" cause issues?  For instance, could you have two distinct code paths that require the same file be utilized?  I didn't see a test of this case is why I ask.  What happens if the two code paths utilize distinctly separate parts of the same file?  I would expect that to not cause problems, from a user standpoint.

**Comment by bradcray** [comment]
I believe those cases should work without issues.

**Pull request merged**